src/config.js

```diff
@@ -0,0 +1,1602 @@
+      if (schema.properties == null) { return value }
+
+      let defaultChildSchema = null
+      let allowsAdditionalProperties = true
+      if (isPlainObject(schema.additionalProperties)) {
+        defaultChildSchema = schema.additionalProperties
+      }
+      if (schema.additionalProperties === false) {
+        allowsAdditionalProperties = false
+      }
+
+      const newValue = {}
+      for (let prop in value) {
+        const propValue = value[prop]
+        const childSchema = schema.properties[prop] != null ? schema.properties[prop] : defaultChildSchema
+        if (childSchema != null) {
+          try {
+            newValue[prop] = this.executeSchemaEnforcers(pushKeyPath(keyPath, prop), propValue, childSchema)
+          } catch (error) {
+            console.warn(`Error setting item in object: ${error.message}`)
+          }
+        } else if (allowsAdditionalProperties) {
+          // Just pass through un-schema'd values
+          newValue[prop] = propValue
+        } else {
+          console.warn(`Illegal object key: ${keyPath}.${prop}`)
+        }
+      }
+
+      return newValue
+    }
+  },
+
+  'array': {
+    coerce (keyPath, value, schema) {
+      if (!Array.isArray(value)) { throw new Error(`Validation failed at ${keyPath}, ${JSON.stringify(value)} must be an array`) }
+      const itemSchema = schema.items
+      if (itemSchema != null) {
+        const newValue = []
+        for (let item of Array.from(value)) {
+          try {
+            newValue.push(this.executeSchemaEnforcers(keyPath, item, itemSchema))
+          } catch (error) {
+            console.warn(`Error setting item in array: ${error.message}`)
+          }
+        }
+        return newValue
+      } else {
+        return value
+      }
+    }
+  },
+
+  'color': {
+    coerce (keyPath, value, schema) {
+      const color = Color.parse(value)
+      if (color == null) {
+        throw new Error(`Validation failed at ${keyPath}, ${JSON.stringify(value)} cannot be coerced into a color`)
+      }
+      return color
+    }
+  },
+
+  '*': {
+    coerceMinimumAndMaximum (keyPath, value, schema) {
+      if (typeof value !== 'number') { return value }
+      if ((schema.minimum != null) && (typeof schema.minimum === 'number')) {
+        value = Math.max(value, schema.minimum)
+      }
+      if ((schema.maximum != null) && (typeof schema.maximum === 'number')) {
+        value = Math.min(value, schema.maximum)
+      }
+      return value
+    },
+
+    validateEnum (keyPath, value, schema) {
+      let possibleValues = schema.enum
+
+      if (Array.isArray(possibleValues)) {
+        possibleValues = possibleValues.map(value => {
+          if (value.hasOwnProperty('value')) { return value.value } else { return value }
+        })
+      }
+
+      if ((possibleValues == null) || !Array.isArray(possibleValues) || !possibleValues.length) { return value }
+
+      for (let possibleValue of Array.from(possibleValues)) {
```

Comment:
we already know this is an array (ln 1562) so we don't need the `Array.from()` here